### PR TITLE
Update deployment settings git configuration to support directories that are not a git repo

### PR DESCRIPTION
**Pull request by glena**
Change how the deployment settings configures git repos to support directories that are not git repositories

Fix https://github.com/pulumi/pulumi-service/issues/20675

### CODE REVIEW

File: pkg/cmd/pulumi/deployment_settings_config.go [hunk 1]
```diff
@@ -327,32 +328,98 @@ func newDeploymentSettingsConfigureCmd() *cobra.Command {
...
+	Repo     *git.Repository
+	Head     *plumbing.Reference
+}
+
+func (r *repoLookupImpl) GetRootDirectory(wd string) (string, error) {
+	return filepath.Rel(r.RepoRoot, wd)
+}
+
+func (r *repoLookupImpl) GetBranchName() string {
+	return r.Head.Name().String()
```
Comment (lblackstone) [inline pkg/cmd/pulumi/deployment_settings_config.go:369]: This probably needs a nil check before dereferencing `Head`.

File: pkg/cmd/pulumi/deployment_settings_config.go [hunk 2]
```diff
@@ -327,32 +328,98 @@ func newDeploymentSettingsConfigureCmd() *cobra.Command {
...
+}
+
+func (r *repoLookupImpl) RemoteURL() (string, error) {
+	return gitutil.GetGitRemoteURL(r.Repo, "origin")
```
Comment (lblackstone) [inline pkg/cmd/pulumi/deployment_settings_config.go:373]: This also might need a nil check

File: sdk/go/common/util/gitutil/git.go [hunk 3]
```diff
@@ -862,6 +862,8 @@ func (a *detectGitAccessorsImpl) Stat(name string) (detectGitFileInfo, error) {
 	return os.Stat(name)
 }
 
+var ErrNoGitRepo = errors.New("Git repo not detected")
```
Comment (lblackstone) [inline sdk/go/common/util/gitutil/git.go:865]:
```suggestion
var ErrNoGitRepo = errors.New("git repo not detected")
```

File: pkg/cmd/pulumi/deployment_settings_config.go [hunk 4]
```diff
@@ -327,32 +328,98 @@ func newDeploymentSettingsConfigureCmd() *cobra.Command {
...
 	wd, err := os.Getwd()
 	if err != nil {
 		return err
 	}
 
 	repoRoot, err := gitutil.DetectGitRootDirectory(wd)
-	if err != nil {
-		return err
+	if err != nil && !errors.Is(err, gitutil.ErrNoGitRepo) {
+		return fmt.Errorf("Could not determine the git root directory: %w", err)
```
Comment (lblackstone) [inline pkg/cmd/pulumi/deployment_settings_config.go:407]:
```suggestion
		return fmt.Errorf("could not determine the git root directory: %w", err)
```
nit: error messages usually don't start with a capital letter

File: pkg/cmd/pulumi/deployment_settings_config.go (hunk 1 above)
Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_config.go:372]: I believe it is unlikely if it doesnt return an error but wont hurt to be defensive, done

File: pkg/cmd/pulumi/deployment_settings_config.go (hunk 2 above)
Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_config.go:379]: same

File: pkg/cmd/pulumi/deployment_settings_config.go [hunk 5]
```diff
@@ -327,32 +328,102 @@ func newDeploymentSettingsConfigureCmd() *cobra.Command {
 	return cmd
 }
 
+func newRepoLookup(repoRoot string) (repoLookup, error) {
+	if repoRoot != "" {
+
```
Comment (tgummerer) [inline pkg/cmd/pulumi/deployment_settings_config.go:333]:
```suggestion
```

File: pkg/cmd/pulumi/deployment_settings_config.go [hunk 6]
```diff
@@ -327,32 +328,102 @@ func newDeploymentSettingsConfigureCmd() *cobra.Command {
...
+		repo, err := git.PlainOpen(repoRoot)
+		if err != nil {
+			return nil, err
+		}
```
Comment (tgummerer) [inline pkg/cmd/pulumi/deployment_settings_config.go:337]: Could we do something like the following here and then we wouldn't even need the `detectGitRootDirectory` function?

```suggestion
		repo, err := git.PlainOpen(repoRoot)
		if err == git.ErrRepositoryNotExists {
			return  &noRepoLookupImpl{}, nil
		}
		if err != nil {
			return nil, err
		}
```

File: pkg/cmd/pulumi/deployment_settings_config.go [hunk 7]
```diff
@@ -361,58 +432,65 @@ func configureGit(ctx context.Context, displayOpts *display.Options, be backend.
...
 
-	remoteURL, err := gitutil.GetGitRemoteURL(repo, "origin")
-	if err != nil {
-		return err
-	}
+	vcsInfo, err := gitutil.TryGetVCSInfo(remoteURL)
 
-	if vcsInfo, err := gitutil.TryGetVCSInfo(remoteURL); err == nil {
-		useGiHub := vcsInfo.Kind == gitutil.GitHubHostName
+	var useGiHub bool
```
Comment (tgummerer) [inline pkg/cmd/pulumi/deployment_settings_config.go:468]:
```suggestion
	var useGitHub bool
```

(affects everything below as well of course)

Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_config.go]: done

File: pkg/cmd/pulumi/deployment_settings_config.go (hunk 6 above)
Comment (glena) [reply pkg/cmd/pulumi/deployment_settings_config.go]: much cleaner now, done

**Review by tgummerer — APPROVED** [review]

**Pull request merged**